### PR TITLE
Pickup changes from GFSv16.3, and update 1D-decomposed GTG compatible with UPP

### LINKED ISSUE #644: Make 1D decomposition of GTG compatible with UPP
NCAR GTG group requested to keep GTG 1D decomposition, while UPP has updated to 2D decomposition.
However there were runtime errors. 
It turned out arrays in UPP have x-dimension ista_2l:iend_2u as 0:IM+1 for GFS.

### LINKED ISSUE #654: Sync wafs related changes from GFS V16.3 Implementation
The related PRs are:
https://github.com/NOAA-EMC/UPP/pull/497
https://github.com/NOAA-EMC/UPP/pull/542 (Wrong procedures. Reopen as #543)
https://github.com/NOAA-EMC/UPP/pull/543

### PR DESCRIPTION
**Pull request by YaliMao-NOAA**
- Manually merge UPP changes made to GFS V16.3 branch as described in issue #654

- As described in issue #644 
It's a temporary and straightforward solution.
1. Reshape array inputs of gtg_algo in MDLFLD.f. GTG has its own exch()
2. Change hard code numx=2 overwritable. Set numx=1 in job card when running GTG
3. As a reminder, set mem=300GB in job card when running GTG

### CODE REVIEW

**Comment by WenMeng-NOAA** [comment]
@YaliMao-NOAA Please sync your branch with the top of  the UPP develop.

**Comment by WenMeng-NOAA** [comment]
@YaliMao-NOAA It seems to me a lot of changes in this PR are synced from WAFS related updates from GFS V16.3 implementation?

**Comment by YaliMao-NOAA** [comment]
> @YaliMao-NOAA It seems to me a lot of changes in this PR are synced from WAFS related updates from GFS V16.3 implementation?

Yes, the changes include three GFSv16.3 PRs:
https://github.com/NOAA-EMC/UPP/pull/497
https://github.com/NOAA-EMC/UPP/pull/542
https://github.com/NOAA-EMC/UPP/pull/543

**Comment by WenMeng-NOAA** [comment]
> > @YaliMao-NOAA It seems to me a lot of changes in this PR are synced from WAFS related updates from GFS V16.3 implementation?
> 
> Yes, the changes include three GFSv16.3 PRs: #497 #542 #543

@YaliMao-NOAA Can you open a new UPP issue for syncing changes of WAFS from GFS V16.3 implementation?

**Comment by WenMeng-NOAA** [comment]
@YaliMao-NOAA Can you sync your branch one more time? Thanks!

**Comment by YaliMao-NOAA** [comment]
@WenMeng-NOAA 
I did the synchronization.  I did a test run, and it turned out the master file is bit-wise consistent with the older version. Is it supposed the recent updates won't change the master file?
Thank you.

File: parm/post_avblflds.xml [hunk 1]
```diff
@@ -4074,7 +4074,7 @@
        <scale>3.0</scale>
     </param>
 
-<!-- 467-469 EDR on FD_HEIGHT_LVL -->
+<!-- 467-469 EDR on FD_HEIGHT_LVL (for regional model)-->
```
Comment (WenMeng-NOAA) [inline parm/post_avblflds.xml:4077]: @YaliMao-NOAA Are you saying ID 467-469 are used for regional models such as RRFS?

File: sorc/ncep_post.fd/MDL2STD_P.f [hunk 2]
```diff
@@ -505,7 +447,7 @@ SUBROUTINE MDL2STD_P()
 
          ! Relabel the pressure level to reference levels
 !         IDS = 0
-         IDS = (/ 450,480,464,465,466,518,519,520,521,522,523,524,525,(0,I=14,50) /)
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/MDL2STD_P.f:508]: @YaliMao-NOAA It appears to me output ID 522 is still kept in this routine. Do you intend to remove it?

File: parm/post_avblflds.xml (hunk 1 above)
Comment (YaliMao-NOAA) [reply parm/post_avblflds.xml:4077]: Yes, these turbulence IDs are used for RAP which outputs turbulence on **flight height** levels, different from GFS outputs turbulence on **pressure levels**
So these IDs are used for a different type of vertical levels

File: sorc/ncep_post.fd/MDL2STD_P.f (hunk 2 above)
Comment (YaliMao-NOAA) [reply sorc/ncep_post.fd/MDL2STD_P.f:508]: This is a good point. I am going to consult UK and AWC whether to keep it for GFSv17, incluidng U/V/T too.

File: parm/post_avblflds.xml (hunk 1 above)
Comment (WenMeng-NOAA) [reply parm/post_avblflds.xml:4077]: @YaliMao-NOAA Thanks for explaining.

File: sorc/ncep_post.fd/MDLFLD.f [hunk 3]
```diff
@@ -927,8 +928,10 @@ SUBROUTINE MDLFLD
            (IGET(752)>0).OR.(IGET(754)>0).OR.      &
            (IGET(278)>0).OR.(IGET(264)>0).OR.      &
            (IGET(450)>0).OR.(IGET(480)>0).OR.      &
+           (IGET(479)>0).OR.(IGET(481)>0).OR.      &
            (IGET(774)>0).OR.(IGET(747)>0).OR.      &
            (IGET(464)>0).OR.(IGET(467)>0).OR.      &
+           (IGET(470)>0).OR.(IGET(476)>0).OR.      &
            (IGET(629)>0).OR.(IGET(630)>0).OR.      &
            (IGET(470)>0).OR.                       &
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/MDLFLD.f]: @YaliMao-NOAA You may remove this line since "IGET(470)>0" is available in line 934.

Comment (YaliMao-NOAA) [reply sorc/ncep_post.fd/MDLFLD.f]: Thank you! Good catch.

**Comment by WenMeng-NOAA** [comment]
@YaliMao-NOAA Please verify wafs data file gfs.t00z.wafs.grb2f006 at my test /u/wen.meng/ptmp/gfs.2019083000 on Cactus and let me know if there are any discrepancies with your test.

File: sorc/ncep_post.fd/MDL2STD_P.f (hunk 2 above)
Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/MDL2STD_P.f:508]: @YaliMao-NOAA You might keep ID 522 in this PR and change later if it is needed. In that case, can you change "521,(0,I=10,50)" into "521,522,(0,I=11,50)"?

**Comment by YaliMao-NOAA** [comment]
> @YaliMao-NOAA Please verify wafs data file gfs.t00z.wafs.grb2f006 at my test /u/wen.meng/ptmp/gfs.2019083000 on Cactus and let me know if there are any discrepancies with your test.

I compared your wafs output and mine at /lfs/h2/emc/ptmp/yali.mao/upp/com/gfs.20190830/00. They are not bit-wise consistent but with good value of rpn_corr and rpn_rms. 

Update: sorry I forgot to mention that the master files are bit-wise consistent.

File: sorc/ncep_post.fd/MDL2STD_P.f (hunk 2 above)
Comment (YaliMao-NOAA) [reply sorc/ncep_post.fd/MDL2STD_P.f:508]: @WenMeng-NOAA I removed 522 (for RH_ON_ICAO_STD_SFC) from UPP, and the code was modified accordingly.

**Comment by FernandoAndrade-NOAA** [comment]
Hello, UPP RT tests have completed on Hera and Orion with no changes to results.

**Comment by WenMeng-NOAA** [comment]
> 

To generate gtg product, should I still overwrite UPP/parm/gtg.cong.gfs with UPP/sorc/ncep_post.fd/post_gtg.fd/gtg.config.gfs? @YaliMao-NOAA

**Comment by YaliMao-NOAA** [comment]
> > 
> 
> To generate gtg product, should I still overwrite UPP/parm/gtg.cong.gfs with UPP/sorc/ncep_post.fd/post_gtg.fd/gtg.config.gfs? @YaliMao-NOAA

Good point. Yes, sorc/ncep_post.fd/post_gtg.fd/gtg.config.gfs is the right one to generate GTG

**Comment by WenMeng-NOAA** [comment]
> > > 
> > 
> > 
> > To generate gtg product, should I still overwrite UPP/parm/gtg.cong.gfs with UPP/sorc/ncep_post.fd/post_gtg.fd/gtg.config.gfs? @YaliMao-NOAA
> 
> Good point. Yes, sorc/ncep_post.fd/post_gtg.fd/gtg.config.gfs is the right one to generate GTG

Can you pick up my driver script submit_run_post_gfs_gtg_wcoss2.sh at /u/wen.meng/noscrub/ncep_post/yali/UPP on Cactus to check any difference from yours? @YaliMao-NOAA

**Comment by YaliMao-NOAA** [comment]
I had another test run with a different number of CPUs, my two results are bit-wise consistent.
Yours and mine have most fields with good value of rpn_corr>=0.999 and some fields are bit-wise consistent.

**Comment by WenMeng-NOAA** [comment]
> I had another test run with a different number of CPUs, my two results are bit-wise consistent. Yours and mine have most fields with good value of rpn_corr>=0.999 and some fields are bit-wise consistent.

Can you point me to your run directory? I don't see any reasons which made wafs files not bitwise identical between your test and my test. @YaliMao-NOAA

**Comment by YaliMao-NOAA** [comment]
> > I had another test run with a different number of CPUs, my two results are bit-wise consistent. Yours and mine have most fields with good value of rpn_corr>=0.999 and some fields are bit-wise consistent.
> 
> Can you point me to your run directory? I don't see any reasons which made wafs files not bitwise identical between your test and my test. @YaliMao-NOAA

Source code: /lfs/h2/emc/vpppg/noscrub/yali.mao/git/UPP.v17 on Cactus
Run result: /lfs/h2/emc/ptmp/yali.mao/upp/com/gfs.20190830/00

**Comment by WenMeng-NOAA** [comment]
My new gtg test can duplicate Yali's test with some changes in test configuration.

**Comment by WenMeng-NOAA** [comment]
The UPP RT tests were completed on WCOSS2 with no changed results. This PR is ready for merging.

**Comment by WenMeng-NOAA** [comment]
No baseline update is needed with this PR.

**Review by WenMeng-NOAA — APPROVED** [review]

**Pull request merged**